### PR TITLE
feat: [CDS-74413]: Terragrunt/Terraform BG deployment fails with NPE fix

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
Setting account id as actual account id instead of global account id in case of harness SM, as  manager apis are expecting actual account Id.
tests Done:
Pipeline execution for terraform and terragrunt apply and provision step.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50607)

### CODE REVIEW

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts_JDK11

File: 400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java [hunk 1]
```diff
@@ -967,6 +967,13 @@ private ExecutionResponse executeInternalInherited(ExecutionContext context, Str
       tfVarSource = fetchTfVarS3Source(context);
     }
 
+    boolean encryptDecryptPlanForHarnessSMOnManager =
+        featureFlagService.isEnabled(CDS_TERRAFORM_TERRAGRUNT_PLAN_ENCRYPTION_ON_MANAGER_CG, context.getAccountId())
+        && isHarnessSecretManager(secretManagerConfig);
+    if (secretManagerConfig != null) {
```
Comment (yogesh-chauhan) [inline 400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java:973]: can we put a comment as to why we need this here ?

File: 400-rest/src/main/java/software/wings/sm/states/provision/TerragruntProvisionState.java [hunk 2]
```diff
@@ -374,6 +374,13 @@ private ExecutionResponse executeInternalRegular(ExecutionContext context, Strin
       exportPlanToApplyStep = true;
     }
 
+    boolean encryptDecryptPlanForHarnessSMOnManager =
+        featureFlagService.isEnabled(CDS_TERRAFORM_TERRAGRUNT_PLAN_ENCRYPTION_ON_MANAGER_CG, context.getAccountId())
+        && terragruntStateHelper.isHarnessSecretManager(secretManagerConfig);
+    if (secretManagerConfig != null) {
```
Comment (yogesh-chauhan) [inline 400-rest/src/main/java/software/wings/sm/states/provision/TerragruntProvisionState.java:380]: same: lets add a comment here

**Review by yogesh-chauhan — APPROVED** [review]

**Review by vlica-harness — DISMISSED** [review]

File: 400-rest/src/main/java/software/wings/sm/states/provision/TerragruntProvisionState.java (hunk 2 above)
Comment (sourabh-awashti) [reply 400-rest/src/main/java/software/wings/sm/states/provision/TerragruntProvisionState.java:380]: done

File: 400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java (hunk 1 above)
Comment (sourabh-awashti) [reply 400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java:973]: done

**Comment by sourabh-awashti** [comment]
trigger uts_JDK11

**Review by yogesh-chauhan — APPROVED** [review]

**Review by vlica-harness — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger uts_JDK11

**Comment by sourabh-awashti** [comment]
trigger ut3

**Comment by sourabh-awashti** [comment]
trigger ut5, ut6, ut7, ut8, ut9, ut0, ut1, ut2, ut3

**Comment by sourabh-awashti** [comment]
trigger review

**Comment by sourabh-awashti** [comment]
trigger ut7

**Comment by yogesh-chauhan** [comment]
*[`400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java` line 1355 at r2](https://reviewable.io/reviews/harness/harness-core/50607#-N_YnUvu6M3K_TLobUiG:-N_YnUvu6M3K_TLobUiH:b-xbzbep) ([raw file](https://github.com/harness/harness-core/blob/8946ad3b7780ed00cd46a79030bc212b89745f72/400-rest/src/main/java/software/wings/sm/states/provision/TerraformProvisionState.java#L1355)):*
> ```Java
>     boolean encryptDecryptPlanForHarnessSMOnManager =
>         featureFlagService.isEnabled(CDS_TERRAFORM_TERRAGRUNT_PLAN_ENCRYPTION_ON_MANAGER_CG, context.getAccountId())
>         && isHarnessSecretManager(secretManagerConfig);
> ```

we should revert this check
isHarnessSecretManager(secretManagerConfig) && featureFlagService.isEnabled(CDS_TERRAFORM_TERRAGRUNT_PLAN_ENCRYPTION_ON_MANAGER_CG, context.getAccountId())

That way we do not call FF service unless needed

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50607)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**